### PR TITLE
o/snapstate: validation sets enforcing on update

### PR DESCRIPTION
**Pull request by stolowski**
Enforce validation sets on refreshes.

### CODE REVIEW

File: overlord/snapstate/storehelpers.go [hunk 1]
```diff
@@ -224,14 +233,51 @@ func installInfo(ctx context.Context, st *state.State, name string, revOpts *Rev
...
+	if len(requiredValSets) > 0 {
+		setActionValidationSets(action, requiredValSets)
+		if !requiredRevision.Unset() {
+			action.Revision = requiredRevision
+		}
+	}
+
+	// check if desired revision matches the revision required by validation sets
+	if !requiredRevision.Unset() && !revOpts.Revision.Unset() && revOpts.Revision.N != requiredRevision.N {
+		return store.SnapActionResult{}, fmt.Errorf("cannot install snap %q at requested revision %s, revision %s required by validation sets: %s", name, revOpts.Revision, requiredRevision, strings.Join(requiredValSets, ","))
```
Comment (mardy) [inline overlord/snapstate/storehelpers.go:270]: nitpick: this line could be split

**Review by mardy — APPROVED** [review]
LGTM, just a nitpick you can also ignore :-)

File: overlord/snapstate/storehelpers.go [hunk 2]
```diff
@@ -519,20 +561,53 @@ func refreshCandidates(ctx context.Context, st *state.State, names []string, use
...
+			if err != nil {
+				return err
+			}
+			// if the snap is already at the required revision then skip it from
+			// candidates when auto-refreshing, otherwise error out.
+			if !requiredRevision.Unset() && installed.Revision == requiredRevision {
+				if opts.IsAutoRefresh {
+					return nil
+				}
+				return fmt.Errorf("snap %q is at the required revision %s by validation sets: %s, cannot update", installed.InstanceName, installed.Revision, strings.Join(requiredValsets, ","))
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:603]: I think this should be a noop too? the user is not really in control of this

File: overlord/snapstate/storehelpers.go [hunk 3]
```diff
@@ -419,6 +443,23 @@ func updateToRevisionInfo(st *state.State, snapst *SnapState, revision snap.Revi
...
+	if err != nil {
+		return nil, err
+	}
+	if enforcedSets != nil {
+		requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(naming.Snap(curInfo.InstanceName()))
+		if err != nil {
+			return nil, err
+		}
+		if !requiredRevision.Unset() && revision != requiredRevision {
+			return nil, fmt.Errorf("cannot update snap %q to revision %s, revision %s is required by validation sets: %s", curInfo.InstanceName(), revision, requiredRevision, strings.Join(requiredValsets, ","))
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:456]: same comment about --ignore-validation

File: overlord/snapstate/storehelpers.go [hunk 4]
```diff
@@ -323,6 +324,29 @@ func updateInfo(st *state.State, snapst *SnapState, opts *RevisionOptions, userI
...
+	if err != nil {
+		return nil, err
+	}
+	if enforcedSets != nil {
+		requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(naming.Snap(curInfo.InstanceName()))
+		if err != nil {
+			return nil, err
+		}
+		if !requiredRevision.Unset() && snapst.Current == requiredRevision {
+			return nil, fmt.Errorf("snap %q is at the required revision %s by validation sets: %s, cannot update",
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:337]: I don't think we want a special new error here, we just want to reproduce the behavior if we got back the the current revision from the store which is handled here I think:

https://github.com/snapcore/snapd/blob/master/store/store_action.go#L612

File: overlord/snapstate/refreshhints_test.go [hunk 5]
```diff
@@ -74,8 +75,9 @@ func (r *recordingStore) SnapAction(ctx context.Context, currentSnaps []*store.C
 type refreshHintsTestSuite struct {
 	state *state.State
 
-	store        *recordingStore
-	restoreModel func()
+	store                         *recordingStore
+	restoreModel                  func()
+	restoreEnforcedValidationSets func()
```
Comment (pedronis) [inline overlord/snapstate/refreshhints_test.go:80]: maybe time to switch to use BaseTest for this?

File: overlord/snapstate/snapstate_remove_test.go [hunk 6]
```diff
@@ -1758,6 +1758,14 @@ func (s *validationSetsSuite) SetUpTest(c *C) {
 		snapstate.AutomaticSnapshot = oldAutomaticSnapshot
 	})
```
Comment (pedronis) [inline overlord/snapstate/snapstate_remove_test.go:1759]: I missed where this was located, I understand that we started with remove, but maybe this suite if we want to share only  one, should be defined in snapstate_install_test.go ?

**Review by pedronis — COMMENTED** [review]
some comments about errors and the corner of having already the right revision.

also some comments about the test suites

File: overlord/snapstate/storehelpers.go [hunk 7]
```diff
@@ -323,6 +324,29 @@ func updateInfo(st *state.State, snapst *SnapState, opts *RevisionOptions, userI
...
+				curInfo.InstanceName(), snapst.Current, strings.Join(requiredValsets, ","))
+		}
+
+		if len(requiredValsets) > 0 {
+			setActionValidationSets(action, requiredValsets)
+			// XXX: what about cohort above?
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:343]: same as install we probably want to ignore it?

File: overlord/snapstate/storehelpers.go [hunk 8]
```diff
@@ -519,20 +561,53 @@ func refreshCandidates(ctx context.Context, st *state.State, names []string, use
...
+			// if the snap is already at the required revision then skip it from
+			// candidates when auto-refreshing, otherwise error out.
+			if !requiredRevision.Unset() && installed.Revision == requiredRevision {
+				if opts.IsAutoRefresh {
+					return nil
+				}
+				return fmt.Errorf("snap %q is at the required revision %s by validation sets: %s, cannot update", installed.InstanceName, installed.Revision, strings.Join(requiredValsets, ","))
+			}
+			if !requiredRevision.Unset() {
+				action.Revision = requiredRevision
```
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:606]: we might need to reset cohort in installed in this case?

**Review by pedronis — COMMENTED** [review]
couple comments about cohorts

File: overlord/snapstate/storehelpers.go (hunk 2 above)
Comment (stolowski) [reply overlord/snapstate/storehelpers.go:603]: Indeed, thanks.

**Comment by stolowski** [comment]
Thinking about it more (and looking at the spec) I think I missed some details in the refresh implementation, I think we need to discuss them, I'm marking this blocked till then.

File: overlord/snapstate/refreshhints_test.go [hunk 9]
```diff
@@ -109,14 +112,19 @@ func (s *refreshHintsTestSuite) SetUpTest(c *C) {
...
+		return nil, nil
+	})
+	s.AddCleanup(restoreEnforcedValidationSets)
 }
 
 func (s *refreshHintsTestSuite) TearDownTest(c *C) {
+	s.BaseTest.TearDownTest(c)
 	dirs.SetRootDir("/")
 	snapstate.CanAutoRefresh = nil
 	snapstate.AutoAliases = nil
```
Comment (pedronis) [inline overlord/snapstate/refreshhints_test.go:127]: nitpick: these can also be do with AddCleanup I suppose and drrop the TearDownTest?

**Review by pedronis — COMMENTED** [review]
overall lgtm, wondering if we need to reset the current snaps CohortKey for snaps controlled by validation-sets in more places?

this will need changes once the InstallMany PR has landed

File: overlord/snapstate/storehelpers.go [hunk 10]
```diff
@@ -318,9 +319,35 @@ func updateInfo(st *state.State, snapst *SnapState, opts *RevisionOptions, userI
...
+		if len(requiredValsets) > 0 {
+			setActionValidationSets(action, requiredValsets)
+			if !requiredRevision.Unset() {
+				action.Revision = requiredRevision
+			}
+		}
+	}
+
+	if action.Revision.Unset() {
+		action.CohortKey = opts.CohortKey
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/snapstate/storehelpers.go:351]: Indeed, done.

**Review by pedronis — APPROVED** [review]
thanks

File: overlord/snapstate/storehelpers.go [hunk 11]
```diff
@@ -422,6 +455,36 @@ func updateToRevisionInfo(st *state.State, snapst *SnapState, revision snap.Revi
...
+	}
+	if enforcedSets != nil {
+		requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(naming.Snap(curInfo.InstanceName()))
+		if err != nil {
+			return nil, err
+		}
+		if !requiredRevision.Unset() {
+			if revision != requiredRevision {
+				return nil, fmt.Errorf("cannot update snap %q to revision %s without --ignore-validation, revision %s is required by validation sets: %s",
+					curInfo.InstanceName(), revision, requiredRevision, strings.Join(requiredValsets, ","))
```
Comment (mardy) [inline overlord/snapstate/storehelpers.go:470]: Is it correct to do this even if `requiredValsets` is empty, or should this `if` block be merged together with the next one?

**Review by mardy — APPROVED** [review]
LGTM, with a question.

File: overlord/snapstate/storehelpers.go (hunk 11 above)
Comment (stolowski) [reply overlord/snapstate/storehelpers.go:470]: It is correct, requiredRevision will only be set if there are requiredValsets (in other words, required revision number comes from validation sets that require it). Of course requiredRevision may be unset even if we have requiredValsets here - it depends on whether the validation set assertions have the optional "required: <revision>" attribute for the given snap or not.

**Pull request merged**